File: src/colibri/sources/markdown.py
"""Plain markdown folder content source.

Handles simple markdown folders for plain markdown folders.
Supports optional YAML frontmatter but doesn't require it.
"""

import re
from collections.abc import Iterator
from pathlib import Path

from colibri.sources.base import ContentSource, SourceDocument


class MarkdownFolderSource(ContentSource):
    """Content source for plain markdown folders.

    A general-purpose source for indexing a folder of markdown files.

    Features:
    - Optional YAML frontmatter (graceful fallback if missing)
    - Relative path links (standard markdown style)
    - Recursive or flat folder scanning
    - Title extraction from filename if not in frontmatter

    Example:
        source = MarkdownFolderSource(
            folder_path=Path("~/Documents/Notes"),
            recursive=True,
            name="My Notes"
        )

        for doc_path in source.list_documents():
            doc = source.read_document(doc_path)
            print(f"{doc.title}")
    """

    source_type = "markdown"

    def __init__(
        self,
        folder_path: str | Path,
        recursive: bool = True,
        name: str | None = None,
        extensions: tuple[str, ...] = (".md",),
        exclude_paths: tuple[str, ...] = (),
    ):
        """Initialize markdown folder source.

        Args:
            folder_path: Path to the markdown folder
            recursive: Whether to scan subfolders (default: True)
            name: Display name for this source (default: folder name)
            extensions: File extensions to index (default: (".md",))
            exclude_paths: Absolute paths to exclude (nested source roots)
        """
        self._folder_path = Path(folder_path).expanduser().resolve()
        self._recursive = recursive
        self._name = name or self._folder_path.name
        self._extensions = extensions
        self._exclude_paths = tuple(
            str(Path(p).resolve()) for p in exclude_paths
        )

    @property
    def name(self) -> str:
        """Source display name."""
        return self._name

    @property
    def root_path(self) -> Path:
        """Folder root path."""
        return self._folder_path

    @property
    def recursive(self) -> bool:
        """Whether subfolders are scanned."""
        return self._recursive

    def list_documents(self, folders: list[str] | None = None) -> Iterator[Path]:
        """Yield paths to all matching files.

        Args:
            folders: Optional subfolder filter (scans these subfolders only)

        Yields:
            Paths relative to folder root
        """
        seen: set[Path] = set()

        def _glob_base(base: Path) -> Iterator[Path]:
            for ext in self._extensions:
                pattern = f"**/*{ext}" if self._recursive else f"*{ext}"
                for fp in base.glob(pattern):
                    rel = fp.relative_to(self._folder_path)
                    if rel in seen:
                        continue
                    if any(part.startswith(".") for part in rel.parts):
                        continue
                    # Skip files under excluded (nested) source paths
                    if self._exclude_paths:
                        abs_str = str(fp.resolve())
                        if any(abs_str.startswith(ex + "/") or abs_str == ex
                               for ex in self._exclude_paths):
                            continue
                    seen.add(rel)
                    yield rel

        if folders:
            for folder in folders:
                folder_path = self._folder_path / folder
                if folder_path.exists():
                    yield from _glob_base(folder_path)
        else:
            yield from _glob_base(self._folder_path)

    def read_document(self, path: Path) -> SourceDocument:
        """Read and parse a document file.

        Handles markdown (.md) with optional YAML frontmatter, and
        YAML/YML files (e.g. OpenAPI specs) as plain text.

        Args:
            path: Path relative to folder root

        Returns:
            SourceDocument with available metadata

        Raises:
            FileNotFoundError: If file doesn't exist
        """
        full_path = self._folder_path / path

        if not full_path.exists():
            raise FileNotFoundError(f"Document not found: {path}")

        suffix = full_path.suffix.lower()

        if suffix in (".yaml", ".yml"):
            return self._read_yaml(path, full_path)

        return self._read_markdown(path, full_path)

    def _read_markdown(self, path: Path, full_path: Path) -> SourceDocument:
        """Read a markdown file with optional frontmatter."""
        content = full_path.read_text(encoding="utf-8")

        # Try to parse frontmatter
        metadata: dict = {}
        doc_content = content

        if content.startswith("---"):
            try:
                import frontmatter

                post = frontmatter.loads(content)
                metadata = dict(post.metadata)
                doc_content = post.content
            except Exception:
                pass

        # Extract title: frontmatter > first H1 > filename
        title = metadata.get("title")
        if not title:
            title = self._extract_first_heading(doc_content)
        if not title:
            title = self._title_from_filename(path.stem)

        folder = path.parts[0] if len(path.parts) > 1 else ""

        tags = metadata.get("tags", [])
        if isinstance(tags, str):
            tags = [t.strip() for t in tags.split(",")]

        return SourceDocument(
            path=path,
            content=doc_content,
            title=title,
            doc_type=metadata.get("type", "note"),
            source_name=self.name,
            source_type=self.source_type,
            folder=folder,
            metadata=metadata,
            tags=tags,
        )

    def _read_yaml(self, path: Path, full_path: Path) -> SourceDocument:
        """Read a YAML file (e.g. OpenAPI spec) as a document."""
        content = full_path.read_text(encoding="utf-8")

        # Try to extract title from OpenAPI info.title
        title = self._extract_yaml_title(content)
        if not title:
            title = self._title_from_filename(path.stem)

        folder = path.parts[0] if len(path.parts) > 1 else ""

        return SourceDocument(
            path=path,
            content=content,
            title=title,
            doc_type="note",
            source_name=self.name,
            source_type=self.source_type,
            folder=folder,
            metadata={},
            tags=[],
        )

    @staticmethod
    def _extract_yaml_title(content: str) -> str | None:
        """Extract title from OpenAPI info block or top-level title key."""
        # Match `title:` inside an `info:` block (OpenAPI convention)
        info_pat = r"^info:\s*\n(?:[ \t]+\S.*\n)*?[ \t]+title:\s*(.+)"
        match = re.search(info_pat, content, re.MULTILINE)
        if match:
            return match.group(1).strip().strip("'\"")
        # Fallback: top-level `title:` key
        match = re.search(r"^title:\s*(.+)", content, re.MULTILINE)
        if match:
            return match.group(1).strip().strip("'\"")
        return None

    def resolve_link(self, link: str, from_doc: Path | None = None) -> Path | None:
        """Resolve a relative markdown link to a document path.

        Handles standard markdown links:
        - ./sibling.md
        - ../parent/doc.md
        - subfolder/doc.md

        Args:
            link: Relative path to resolve
            from_doc: Document containing the link (for relative resolution)

        Returns:
            Path relative to folder root, or None if not found
        """
        # Clean up link (remove any URL encoding, anchors)
        clean_link = link.split("#")[0]  # Remove anchor
        clean_link = clean_link.split("?")[0]  # Remove query

        if not clean_link:
            return None

        # Resolve relative to source document or root
        base_dir = (self._folder_path / from_doc).parent if from_doc else self._folder_path
        resolved = (base_dir / clean_link).resolve()

        # Security: ensure resolved path is within folder
        try:
            resolved.relative_to(self._folder_path)
        except ValueError:
            return None

        # Check if file exists
        if not resolved.exists():
            # Try adding .md extension
            if not clean_link.endswith(".md"):
                resolved_with_ext = resolved.with_suffix(".md")
                if resolved_with_ext.exists():
                    return resolved_with_ext.relative_to(self._folder_path)
            return None

        return resolved.relative_to(self._folder_path)

    def _extract_first_heading(self, content: str) -> str | None:
        """Extract the first H1 heading from content."""
        # Match # Heading at start of line
        match = re.search(r"^#\s+(.+)$", content, re.MULTILINE)
        if match:
            return match.group(1).strip()
        return None

    def _title_from_filename(self, stem: str) -> str:
        """Convert filename stem to readable title."""
        # Replace common separators with spaces
        title = stem.replace("-", " ").replace("_", " ")
        # Title case
        return title.title()

File: src/sources/markdown.rs
//! Plain markdown folder content source.
//!
//! Handles simple markdown folders with optional YAML frontmatter.
//! Mirrors the Python `MarkdownFolderSource`.

use std::collections::HashSet;
use std::path::{Path, PathBuf};

use regex::Regex;

use crate::error::ColibriError;
use crate::sources::{ContentSource, SourceDocument};

/// Content source for plain markdown folders.
pub struct MarkdownFolderSource {
    folder_path: PathBuf,
    recursive: bool,
    display_name: String,
    extensions: Vec<String>,
    exclude_paths: Vec<String>,
}

impl MarkdownFolderSource {
    pub fn new(
        folder_path: impl Into<PathBuf>,
        recursive: bool,
        name: Option<String>,
        extensions: Vec<String>,
        exclude_paths: Vec<String>,
    ) -> Self {
        let folder_path = folder_path.into();
        let folder_path = if folder_path.starts_with("~") {
            if let Some(home) = dirs::home_dir() {
                home.join(folder_path.strip_prefix("~").unwrap())
            } else {
                folder_path
            }
        } else {
            folder_path
        };
        let folder_path = folder_path
            .canonicalize()
            .unwrap_or_else(|_| folder_path.clone());

        let display_name = name.unwrap_or_else(|| {
            folder_path
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
                .to_string()
        });

        // Resolve exclude paths to absolute
        let exclude_paths = exclude_paths
            .into_iter()
            .map(|p| {
                let pb = PathBuf::from(&p);
                pb.canonicalize()
                    .unwrap_or(pb)
                    .to_string_lossy()
                    .into_owned()
            })
            .collect();

        Self {
            folder_path,
            recursive,
            display_name,
            extensions,
            exclude_paths,
        }
    }
}

impl ContentSource for MarkdownFolderSource {
    fn name(&self) -> &str {
        &self.display_name
    }

    fn root_path(&self) -> &Path {
        &self.folder_path
    }

    fn source_type(&self) -> &str {
        "markdown"
    }

    fn list_documents(&self) -> Vec<PathBuf> {
        let mut seen = HashSet::new();
        let mut results = Vec::new();

        if !self.folder_path.exists() {
            return results;
        }

        for ext in &self.extensions {
            let pattern = if self.recursive {
                format!("{}/**/*{ext}", self.folder_path.display())
            } else {
                format!("{}/*{ext}", self.folder_path.display())
            };

            let entries = match glob::glob(&pattern) {
                Ok(entries) => entries,
                Err(_) => continue,
            };

            for entry in entries.flatten() {
                let rel = match entry.strip_prefix(&self.folder_path) {
                    Ok(r) => r.to_path_buf(),
                    Err(_) => continue,
                };

                // Skip duplicates
                if seen.contains(&rel) {
                    continue;
                }

                // Skip hidden files/directories
                if rel
                    .components()
                    .any(|c| c.as_os_str().to_str().is_some_and(|s| s.starts_with('.')))
                {
                    continue;
                }

                // Skip files under excluded paths
                if !self.exclude_paths.is_empty() {
                    let abs_str = entry.to_string_lossy();
                    if self.exclude_paths.iter().any(|ex| {
                        abs_str.starts_with(&format!("{ex}/")) || abs_str.as_ref() == ex.as_str()
                    }) {
                        continue;
                    }
                }

                seen.insert(rel.clone());
                results.push(rel);
            }
        }

        results.sort();
        results
    }

    fn read_document(&self, path: &Path) -> Result<SourceDocument, ColibriError> {
        let full_path = self.folder_path.join(path);

        if !full_path.exists() {
            return Err(ColibriError::Source(format!(
                "Document not found: {}",
                path.display()
            )));
        }

        let suffix = full_path.extension().and_then(|e| e.to_str()).unwrap_or("");

        match suffix {
            "yaml" | "yml" => self.read_yaml(path, &full_path),
            _ => self.read_markdown(path, &full_path),
        }
    }
}

impl MarkdownFolderSource {
    /// Read a markdown file with optional YAML frontmatter.
    fn read_markdown(&self, path: &Path, full_path: &Path) -> Result<SourceDocument, ColibriError> {
        let content = read_lossy(full_path)?;

        let (metadata, doc_content) = parse_frontmatter(&content);

        // Extract title: frontmatter > first H1 > filename
        let title = metadata
            .get("title")
            .and_then(|v| v.as_str())
            .map(String::from)
            .or_else(|| extract_first_heading(&doc_content))
            .unwrap_or_else(|| title_from_filename(path));

        let folder = path
            .components()
            .next()
            .and_then(|c| c.as_os_str().to_str())
            .filter(|_| path.components().count() > 1)
            .unwrap_or("")
            .to_string();

        // Extract tags
        let tags = extract_tags(&metadata);

        // Extract doc_type from frontmatter "type" field
        let doc_type = metadata
            .get("type")
            .and_then(|v| v.as_str())
            .unwrap_or("note")
            .to_string();

        Ok(SourceDocument {
            path: path.to_path_buf(),
            content: doc_content,
            title,
            doc_type,
            source_name: self.display_name.clone(),
            source_type: "markdown".into(),
            folder,
            metadata,
            tags,
        })
    }

    /// Read a YAML file (e.g., OpenAPI spec) as a document.
    fn read_yaml(&self, path: &Path, full_path: &Path) -> Result<SourceDocument, ColibriError> {
        let content = read_lossy(full_path)?;

        let title = extract_yaml_title(&content).unwrap_or_else(|| title_from_filename(path));

        let folder = path
            .components()
            .next()
            .and_then(|c| c.as_os_str().to_str())
            .filter(|_| path.components().count() > 1)
            .unwrap_or("")
            .to_string();

        Ok(SourceDocument {
            path: path.to_path_buf(),
            content,
            title,
            doc_type: "note".into(),
            source_name: self.display_name.clone(),
            source_type: "markdown".into(),
            folder,
            metadata: serde_json::Map::new(),
            tags: vec![],
        })
    }
}

/// Parse YAML frontmatter delimited by `---`.
///
/// Returns `(metadata, content_without_frontmatter)`.
fn parse_frontmatter(text: &str) -> (serde_json::Map<String, serde_json::Value>, String) {
    let empty = (serde_json::Map::new(), text.to_string());

    if !text.starts_with("---") {
        return empty;
    }

    // Find closing ---
    let rest = &text[3..];
    let end = match rest.find("\n---") {
        Some(pos) => pos,
        None => return empty,
    };

    let yaml_str = &rest[..end];
    let content_start = 3 + end + 4; // skip opening ---, yaml, \n---
    let doc_content = if content_start < text.len() {
        text[content_start..].trim_start_matches('\n').to_string()
    } else {
        String::new()
    };

    // Parse YAML into a JSON map
    match serde_yaml::from_str::<serde_json::Value>(yaml_str) {
        Ok(serde_json::Value::Object(map)) => (map, doc_content),
        _ => (serde_json::Map::new(), text.to_string()),
    }
}

/// Extract the first H1 heading from markdown content.
fn extract_first_heading(content: &str) -> Option<String> {
    let re = Regex::new(r"(?m)^#\s+(.+)$").ok()?;
    re.captures(content).map(|cap| cap[1].trim().to_string())
}

/// Convert filename stem to a readable title.
fn title_from_filename(path: &Path) -> String {
    let stem = path
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("Untitled");
    stem.replace(['-', '_'], " ")
}

/// Extract tags from frontmatter metadata.
fn extract_tags(metadata: &serde_json::Map<String, serde_json::Value>) -> Vec<String> {
    match metadata.get("tags") {
        Some(serde_json::Value::Array(arr)) => arr
            .iter()
            .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
            .filter(|s| !s.is_empty())
            .collect(),
        Some(serde_json::Value::String(s)) => s
            .split(',')
            .map(|t| t.trim().to_string())
            .filter(|t| !t.is_empty())
            .collect(),
        _ => vec![],
    }
}

/// Read a file as UTF-8, replacing invalid bytes with the Unicode replacement character.
fn read_lossy(path: &Path) -> Result<String, ColibriError> {
    let bytes = std::fs::read(path).map_err(|e| {
        ColibriError::Source(format!("Failed to read {}: {e}", path.display()))
    })?;
    Ok(String::from_utf8_lossy(&bytes).into_owned())
}

/// Extract title from a YAML document (e.g., OpenAPI spec).
fn extract_yaml_title(content: &str) -> Option<String> {
    // Try info.title (OpenAPI convention)
    let info_re = Regex::new(r"(?m)^info:\s*\n(?:[ \t]+\S.*\n)*?[ \t]+title:\s*(.+)").ok()?;
    if let Some(cap) = info_re.captures(content) {
        let title = cap[1].trim().trim_matches(|c| c == '\'' || c == '"');
        if !title.is_empty() {
            return Some(title.to_string());
        }
    }

    // Fallback: top-level title key
    let title_re = Regex::new(r"(?m)^title:\s*(.+)").ok()?;
    title_re.captures(content).map(|cap| {
        cap[1]
            .trim()
            .trim_matches(|c| c == '\'' || c == '"')
            .to_string()
    })
}
